Support deep links to encode content via web links

TonSilver opened the pull request:
Recently, I wanted to store QR codes in my notes as text instead of images, and that got me thinking about this feature in BinaryEye.

I added support for opening `barcode://encode` via an HTTP deep link, as well as an action to copy the deep link to the clipboard. I still need an icon for this action — any suggestions on how to create one would be helpful.

Also, please take a look at the intent filter for the `EncodeBinaryEye` host — maybe there’s a better way to name it?

p.s. Translation i'll do later (with ChatGPT ^_^)

п.с. Use cases where I now use this feature:
* Storing discount cards for grocery stores.
* QR code for fitness club.

TonSilver commented on app/src/main/kotlin/de/markusfisch/android/binaryeye/fragment/DecodeFragment.kt:
```diff
@@ -448,6 +449,12 @@ class DecodeFragment : Fragment() {
 				true
 			}
 
+			R.id.copy_as_deeplink -> {
```
Need to add the same action to BarcodeFragment

markusfisch commented:
This is great! 😄  And very well done! This has to be one of the cleanest PRs I've seen in a long time! 👍 

> …as well as an action to copy the deep link to the clipboard.

Haven't thought about this - I really like that idea!

> I still need an icon for this action — any suggestions on how to create one would be helpful.

How about the `link` icon? Not perfect either, but maybe close enough?
<img width="151" alt="image" src="https://github.com/user-attachments/assets/3d07fc89-da96-4762-b882-9956cc0ecbb2" />

> Also, please take a look at the intent filter for the EncodeBinaryEye host…

I would maybe suggest `encode` because it's shorter. And it would also allow me to offer a fallback website for encoding barcodes at this address. This would then not be tied to Binary Eye, of course.

> p.s. Translation i'll do later (with ChatGPT ^_^)

Yes, I know, translations always take a lot of time, and I would do them with ChatGPT, too. I can add these, too.

TonSilver commented:
I'll do the rest later. In 24 hours.

Could you help me how to get this icon as a xml resource? Where it is located?

markusfisch commented:
> I'll do the rest later. In 24 hours.

That would be nice! But no need to hurry.

> Could you help me how to get this icon as a xml resource? Where it is located?

Sorry, of course, I should've attached a link instead of a screenshot:

https://fonts.google.com/icons?selected=Material+Symbols+Outlined:link:FILL@0;wght@400;GRAD@0;opsz@24&icon.query=link&icon.size=24&icon.color=%23e3e3e3

It's part of Googles' Material icons, which are also available from within Android Studio: right click the `res` folder, choose "New" -> "Vector Asset", check "Clip art" for Asset type and press the button at "Clip art:". There you can enter "link", which was what I initally did 😉

TonSilver commented:
@markusfisch, I’m done :)

It was an interesting task — the translation part, I mean. Hopefully it's correct. I used a diff (±2 lines) with 25 languages and sent it to GPT. It patched my patch with a script, so I just applied the result. Then I asked it to double-check itself (in different chats). Pretty cool!

By the way, I also added a “Copy as deeplink” feature to BarcodeFragment. That led to some refactoring, and I added a few tests (and `robolectric` library, for `android.net.Uri`).

So, if you want to split this pull request, revert any changes I made, or tweak any part of the code — just let me know!

markusfisch commented:
Thank you very much! 😄  The refactorings are very reasonable, and the translations look good, too 👍 

At first I was a bit put off by the inclusion of `robolectric` (I'm very careful with dependencies 😉 ), but it really makes sense here. I actually didn't know `robolectric` yet, but I'll probably use it in other projects as well, so thanks for bringing it to my attention.

And the translation dance with ChatGPT sounds quite interesting indeed! I am, too, still trying to find the best/easiest way to do this, and also wrote a script for https://github.com/markusfisch/BinaryEye/commit/fa38866669e640a2529a63f3dfaef4e11389d107.

TonSilver commented:
> At first I was a bit put off by the inclusion of robolectric

I’m not a fan of bloated libraries either, but this particular one is our go-to tool across all projects—over 100 of them, with hundreds of modules. It’s very useful for writing even tests that need to handle different behaviors across Android SDK versions.

TonSilver commented:
> also wrote a script for

Me too 😄

TonSilver commented:
I mean i'm writing script right now, for fun))

TonSilver commented:
Interesting task. LLMs make translation incredibly easy!

markusfisch commented:
> I’m not a fan of bloated libraries either, but this particular one is our go-to tool across all projects—over 100 of them, with hundreds of modules. It’s very useful for writing even tests that need to handle different behaviors across Android SDK versions.

Yes, I can totally see how useful it is! I usually painfully mock all the Android things or fall back to connected device tests, which aren't too convinient, of course. So I will definitely try it in some of my projects now!

> I mean i'm writing script right now, for fun))

Just for the fun of it, here's my little script that I used to add the missing translations (using the OpenAI API):

```
#!/usr/bin/env bash

ask() {
	local RESPONSE
	RESPONSE=$(curl -s -X POST 'https://api.openai.com/v1/chat/completions' \
		-H "Content-Type: application/json" \
		-H "Authorization: Bearer ${OPENAI_API_KEY}" \
		-d "$(jq -n \
			--arg model 'gpt-3.5-turbo' \
			--arg prompt "$1" \
			'{
				"model": $model,
				"messages": [
					{
						"role": "user",
						"content": $prompt
					}
				],
				"temperature": 0.7
			}'
		)"
	) || return $?

	local CONTENT
	CONTENT=$(echo "$RESPONSE" | \
		jq -r '.choices[0].message.content // empty' \
	)

	if [ -z "$CONTENT" ] || [ "$CONTENT" == 'null' ]; then
		echo "error: $RESPONSE" >&2
		return 1
	fi

	echo "$CONTENT"
}

readonly LF=$'\n'
grep '<string' app/src/main/res/values/strings.xml | while read -r STRING
do
	grep "$STRING" app/src/main/res/values-*/str* | while read -r
	do
		FILE=${REPLY%%:*}
		CC=${FILE#*-}
		CC=${CC%/*}
		PROMPT="Please translate the following string resource into the language of the country code $CC:$LF$LF$STRING${LF}${LF}IMPORTANT: leave the surrounding XML tags unmodified and include them in the answer."
		RESULT=$(ask "$PROMPT") || {
			echo "error: $RESULT" >&2
			exit 1
		}
		RESULT=$(echo "$RESULT" | grep -m1 -o '<string.*string>') || {
			echo "error: result contains no <string/>: [$RESULT]" >&2
			echo "question: [$PROMPT]" >&2
			exit 2
		}
		RESULT=${RESULT%${LF}*}
		sed -i '' "s|${STRING}|${RESULT}|" "$FILE" || exit $?
	done || exit $?
done
```

markusfisch commented:
Not really, I paid $0.12, for 96.122 tokens including testing the script and so on 😉

But of course, it depends on the model and how much you use it.
